packages/worker/src/connection.ts

```diff
@@ -1,29 +1,208 @@
+import * as grpc from '@grpc/grpc-js';
+import { getRelativeTimeout } from '@grpc/grpc-js/build/src/deadline';
+import * as proto from 'protobufjs';
+
 import { IllegalStateError } from '@temporalio/common';
-import { Client, Worker, clientUpdateHeaders, TransportError, clientUpdateApiKey } from '@temporalio/core-bridge';
+import {
+  Client,
+  Worker,
+  clientUpdateHeaders,
+  TransportError,
+  clientUpdateApiKey,
+  clientSendRequest,
+} from '@temporalio/core-bridge';
+import { ConnectionLike, Metadata, CallContext, WorkflowService } from '@temporalio/client';
+import { SymbolBasedInstanceOfError } from '@temporalio/common/lib/type-helpers';
 import { NativeConnectionOptions } from './connection-options';
 import { Runtime } from './runtime';
 
 const updateHeaders = util.promisify(clientUpdateHeaders);
 const updateApiKey = util.promisify(clientUpdateApiKey);
 
+@SymbolBasedInstanceOfError('ServiceError')
+export class ServiceError extends Error {
+  constructor(
+    message: string,
+    public readonly code: grpc.StatusObject['code'],
+    public readonly details: string,
+    public readonly metadata: grpc.Metadata
+  ) {
+    super(message);
+  }
+}
+
+function serviceErrorFromError(err: unknown): ServiceError {
+  if (err instanceof TransportError) {
+    const metadata = new grpc.Metadata();
+    for (const [k, v] of Object.entries(err.metadata ?? {})) {
+      if (typeof v === 'string') {
+        metadata.set(k, v);
+      } else {
+        metadata.set(k, Buffer.from(v));
+      }
+    }
+    const se = new ServiceError(err.message, err.code ?? 0, '', metadata);
+    se.stack = err.stack;
+    return se;
+  }
+  if (err instanceof Error) {
+    const se = new ServiceError(err.message, 0, '', new grpc.Metadata());
+    se.stack = err.stack;
+    return se;
+  }
+  return new ServiceError(`Unknown error: ${err}`, 0, '', new grpc.Metadata());
+}
+
 /**
  * A Native Connection object that delegates calls to the Rust Core binary extension.
  *
  * A Worker must use this class to connect to the server.
  *
- * Do not confuse this connection class with `@temporalio/client`'s Connection.
+ * This class can be used to power `@temporalio/client`'s Client objects.
  */
-export class NativeConnection {
+export class NativeConnection implements ConnectionLike {
   /**
    * referenceHolders is used internally by the framework, it can be accessed with `extractReferenceHolders` (below)
    */
   private readonly referenceHolders = new Set<Worker>();
 
+  public readonly workflowService: WorkflowService;
+  readonly callContextStorage = new AsyncLocalStorage<CallContext>();
+
   /**
    * nativeClient is intentionally left private, framework code can access it with `extractNativeClient` (below)
    */
-  protected constructor(private nativeClient: Client) {}
+  protected constructor(private nativeClient: Client) {
+    this.workflowService = WorkflowService.create(this.sendRequest.bind(this), false, false);
```

Comment:
Definitely not required for this PR, but any reason we should _not_ also implement other services over `NativeConnection` at some later point?

Reply:
We can definitely do that later, I just wanted to limit the scope for this MVP. I will open an issue.